Extend docs on how to write a custom devServer (#5211)
cypress-io/cypress-documentation: docs/guides/component-testing/component-framework-configuration.mdx

Below are more advanced configuration options you can customize to fit your
project.

## Custom Index File

By default, Cypress renders your components into an HTML file located at
`cypress/support/component-index.html`.

The index file allows you to add in global assets, such as styles, fonts, and
external scripts.

You can provide an alternative path to the file using the `indexHtmlFile` option
in the [component config](/guides/references/configuration#component) options:

```js
{
  component: {
    devServer,
    indexHtmlFile: '/custom/path/to/component-index.html'
  }
}
```

## Custom Dev Server

A custom function can be passed into the `devServer` option, which allows the
use of build systems not provided by Cypress out of the box. These can be from
the Cypress community, preview builds not included with the app, or a custom one
you create.

The function's signature takes in a object with the following properties as its
only parameter and needs to resolve an object containing the port of your dev
server.

```ts

> interface DevServerOptions {

Reviewer: Nice - we should probably expose this via `Cypress.DevServerOptions` in the future to help guide third parties.

  specs: Cypress.Spec[]
  cypressConfig: Cypress.PluginConfigOptions
  devServerEvents: NodeJS.EventEmitter
}
```

:::cypress-config-example

```ts
{
  component: {
    async devServer({specs, cypressConfig, devServerEvents}: DevServerOptions) {
      const port = await startDevServer(specs, cypressConfig, devServerEvents)
      return {

>         port: port

Reviewer: Should we include `close` here, maybe

```js
return {
  port,
  close: () => { /* optional code to run when the dev server is closed */ }
}
```
?

Author: I skipped it since in the webpack dev-server it was mentioned it's only for unit testing: https://github.com/cypress-io/cypress/blob/develop/npm/webpack-dev-server/src/devServer.ts#L75
I'm not sure if this is stable enough to be used as a public api?

Author: Hmm, close seems required in the type-definitions, so I'll add it after all.

Below are more advanced configuration options you can customize to fit your
project.

## Custom Index File

By default, Cypress renders your components into an HTML file located at
`cypress/support/component-index.html`.

The index file allows you to add in global assets, such as styles, fonts, and
external scripts.

You can provide an alternative path to the file using the `indexHtmlFile` option
in the [component config](/guides/references/configuration#component) options:

```js
{
  component: {
    devServer,
    indexHtmlFile: '/custom/path/to/component-index.html'
  }
}
```

## Custom Dev Server

A custom function can be passed into the `devServer` option, which allows the
use of build systems not provided by Cypress out of the box. These can be from
the Cypress community, preview builds not included with the app, or a custom one
you create.

> The function's signature takes in a object with the following properties as its

Reviewer 2: should say "takes in `an` object"

  app.use(cypressConfig.devServerPublicPathRoute, express.static(bundleDir))

  app.listen(port)
}
```

For a real-world example, you can refer to
[this loader](https://github.com/cypress-io/cypress/blob/466155c2125476374d9f9549530f67d0c6354a41/npm/vite-dev-server/src/plugins/cypress.ts#L82-L92)
used by the vite-devserver.

The client script must retrieve information on the currently active test from
the Cypress instance of the parent frame and load the corresponding bundle. If a
support file is defined, it should be injected at the top of your test bundle or
loaded before the test script.

```ts
const CypressInstance = (window.Cypress = parent.Cypress)
const devServerPublicPathRoute = CypressInstance.config(
  'devServerPublicPathRoute'
)

const imports = []

// If you do not bundle your support file along with the tests,
// you need to add a separate import statement for the support file.
if (CypressInstance.config('supportFile')) {
  const absolute = CypressInstance.config('supportFile')
  const relative = absolute.replace(CypressInstance.config('projectRoot'), '')
  const relativeUrl = `${devServerPublicPathRoute}${relative}`

>   imports.push({

Author: I had a closer look into the load mechanism, and this format causes the script to be loaded twice:
https://github.com/cypress-io/cypress/blob/7a33f5c1a87f5a3aa3b8da575b39e42f93bfd5db/packages/driver/src/cypress/script_utils.ts#L6
First it will be fetched in the `fetchScript` function, and later in `evalScripts` the load-function is triggered, wich causes the import to fetch the contents again.

It's propably better to just pass the importFunctions and use the alternative path here: https://github.com/cypress-io/cypress/blob/7a33f5c1a87f5a3aa3b8da575b39e42f93bfd5db/packages/driver/src/cypress/script_utils.ts#L49

Author: This propably should be fixed for the vite-dev-server as well

:::

Any requests triggered during a test using the `devServerPublicPathRoute` as

> defined in the `cypressConfig` will be forwarded to your server. The

Reviewer: Might be worth including you can access the user's HTML file via `options.cypressConfig.indexHtmlFile` in your custom integration.

> Any requests triggered during a test using the `devServerPublicPathRoute` as
> defined in the `cypressConfig` will be forwarded to your server. The
> cypress-runner will trigger a request for an index.html when a test is started.
> Your server needs to send the custom index.html file after injecting a script to
> load the actual test. If you do not bundle your support file along with the
> tests, you need to add a separate import statement for the support file.

Reviewer: ```suggestion
Any requests triggered during a test using the `devServerPublicPathRoute` as
defined in the `cypressConfig` will be forwarded to your server. Cypress
will trigger a request for `component-index.html` when a test is started.
Your server needs to reply with a custom `index.html` file after injecting a script to
load the actual test. If you do not bundle your support file along with the
tests, you need to add a separate import statement for the support file.
```

Might be worth linking to an example in the Cypress code base, so third parties have something to reference. What do you think?

Author: Sounds good, I wasn't sure if this is good practice for the docs.

```ts
const CypressInstance = (window.Cypress = parent.Cypress)
const { absolute, relative } = CypressInstance.spec
const relativeUrl = `${CypressInstance.config(
  'devServerPublicPathRoute'
)}/${relative}`

// load the testfile - you can extend the load function to also load css
CypressInstance.onSpecWindow(window, [
  { absolute, relative, relativeUrl, load: () => import(relativeUrl) },
])

// then start the test process
CypressInstance.action('app:window:before:load', window)
```

> The `devServerEvents` event emitter should be used to notify cypress about
> finished builds by emitting the `dev-server:compile:success`-event and to listen
> for the `dev-server:specs:changed`-event that will notify you about changed
> entry points.

Reviewer: ```suggestion
The `devServerEvents` event emitter should be used to notify cypress about
finished builds by emitting a `dev-server:compile:success` event and to listen
for the `dev-server:specs:changed` event that will notify you about changed
entry points.
```

the actual test.

```ts
function createServer(cypressConfig, bundleDir, port = 1234) {
  const app = express()

  // read kickstart script - see below for an example
  const clientScript = readFileSync(
    path.join(__dirname, './client-script.js'),
    'utf8'
  )

  app.get(
    cypressConfig.devServerPublicPathRoute + '/index.html',
    async (_req, res) => {
      // read custom index.html file
      const html = await fs.readFile(
        path.join(cypressConfig.repoRoot, cypressConfig.indexHtmlFile),
        { encoding: 'utf8' }
      )

      // inject kickstart-script
      const output = html.replace(
        '</head>',
        `<script type="module">${clientScript}</script></head>`
      )
      res.send(output)
    }
  )

  // you need to establish some url-to-path-mapping, if your bundler outputs
  // the full directory structure you can map this one to one
  app.use(cypressConfig.devServerPublicPathRoute, express.static(bundleDir))

  app.listen(port)
}
```

For a real-world example, you can refer to
[this loader](https://github.com/cypress-io/cypress/blob/466155c2125476374d9f9549530f67d0c6354a41/npm/vite-dev-server/src/plugins/cypress.ts#L82-L92)

> used by the vite-devserver.

Reviewer: ```suggestion
used by the Vite Dev Server.
```

The client script must retrieve information on the currently active test from
the Cypress instance of the parent frame and load the corresponding bundle. If a
support file is defined, it should be injected at the top of your test bundle or
loaded before the test script.

```ts
const CypressInstance = (window.Cypress = parent.Cypress)
const devServerPublicPathRoute = CypressInstance.config(
  'devServerPublicPathRoute'
)

let importPromise = Promise.resolve()

// If you do not bundle your support file along with the tests,
// you need to add a separate import statement for the support file.
const supportFilePath = CypressInstance.config('supportFile')
if (supportFilePath) {
  const relative = supportFilePath.replace(
    CypressInstance.config('projectRoot'),
    ''
  )
  importPromise = importPromise.then(
    () => import(`${devServerPublicPathRoute}${relative}`)
  )
}

> // load the testfile - you can extend the load function to also load css

Reviewer: ```suggestion
// load the spec - you can extend the load function to also load css
```

Reviewer: Just to clarify:

Spec - a file on your machine
Suite - a `describe`
Test - an `it`
